enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,756 @@
+
+### Non-Goals
+
+* Setting up or configuring any part of Cluster API, this is handled in a separate enhancement
+* Modifying any of the existing deployments of Machine API within clusters
+* Deleting or modifying any existing resources within clusters (until a user chooses to do so)
+* Force creation of new Machine API machines when new Cluster API MachineSets are being created[^4]
+* Port any feature from Cluster API into Machine API
+
+[^4]: We want to promote, over time, a reduction in the number of Machine API Machines, as such,
+any resources created directly in Cluster API, will not be reflected into Machine API, unless they have an owner,
+and, that owner is already reflected into Machine API.
+
+## Proposal
+
+We will implement a two-way synchronisation controller that synchronises Machines and related resources between the Machine API and Cluster API equivalents.
+
+Using an annotation on the Machine API resource, users will be able to choose which API is authoritative.
+Any changes to the non-authoritative resource will be overwritten by the synchronisation controller to match the authoritative resource.
+
+When the API is non-authoritative, the controllers should ignore the resource,
+allowing the authoritative API’s controller to perform the reconciliation actions required.
+
+### Workflow Description
+
+In this workflow, the **cluster admin** is responsible for the infrastructure provisioning for the cluster.
+
+When the cluster admin wishes to migrate a MachineSet from Machine API to Cluster API, the following procedure is required:
+1. Identify the MachineSet to migrate to Cluster API
+1. Use the `oc edit` or a patch command to update the value of the `machines.openshift.io/authoritative-api` annotation to `cluster.x-k8s.io`
+1. The validating webhook verifies that the value of the annotation is valid and that the synchronised generation is up to date, and permits the request
+1. The annotation update is persisted
+1. The Cluster API controller takes over management of the MachineSet going forwards
+
+To migrate back from Cluster API to Machine API, the procedure is the same, however, the value of the annotation should be set to `machine.openshift.io`.
+
+When the conversion cannot proceed for some reason, for example, a feature is in use in Machine API that is not present in Cluster API, the following will happen:
+1. Cluster admin identifies a MachineSet to migrate to Cluster API
+1. Cluster admin uses `oc edit` to update the value of the `machines.openshift.io/authoritative-api` annotation to `cluster.x-k8s.io`
+1. The validating webhook verifies the value of the annotation is valid
+1. The validating webhook observes that no Cluster API equivalent resource has been created, or, the equivalent Cluster API resource is stale
+1. The validating webhook fetches the synchronisation failure condition from the Machine API MachineSet and returns this, along with a reject message, to the Cluster Admin
```

Comment:
Yeah i think there will be conditions, but based on other conversations, we are going to rework the authority handover to be a controller based acceptance rather than webhook based acceptance, so we should revisit this once I've got that sorted